### ISSUE TITLE
[RFC] Support for mock for final class

### ISSUE THREAD

**Issue by Nek-** [phpspec/prophecy#197]
Hello,

I think code should not be change for tests. But currently, to work with prophecy and/or phpspec, that's what people do. (i.e. https://github.com/doctrine/doctrine2/pull/1215)

I think it's time to change that and supports mock for final classes.

What do you think about that ?

**Comment by stof** [comment]
How would you mock final classes ? The generated mock class currently extends the mocked class so that it passes typehints, and so it is not possible for final classes.

AFAICT, mocking final classes would involved defining a mock definition of the class and running all the code for this mock in a PHP process where the original class is never loaded. But this makes things much more complex, because it involves many assumption about the code using the mocking library, and would make the library incompatible with PHPUnit or PhpSpec (meaning we won't do it as the primary goal of creating Prophecy was to fulfill the PhpSpec needs) as they run several tests in the same PHP process.
And it also makes the job of PhpSpec much harder, because it would have to stop using Reflection to get info about the class (as this requires loading the class) and to stop using autoloading to find the class declaration (which would be a huge pain to configure Prophecy as you would have to configure a separate class locator).
Thus, it would also mean than internal classes (i.e. classes defined in C extensions) would not be mockable with such implementation, as well as any class loaded by the test runtime (as they would be loaded in the process).

**Comment by stof** [comment]
Btw, IMO, you should never mock Doctrine. If you mock Doctrine, you will test that your code works against the way you **think** Doctrine works, not against the way Doctrine actually works. And if the way Doctrine works changes in a new version, your testsuite will not tell you about the bug.
Given the complexity of Doctrine, I can tell you that I'm unable to write a reliable mock of its API btw, even if you don't care about make your testsuite useful against Doctrine changes. and I doubt the Doctrine lead dev would be able to do it either.
The general rule of mocking is that you should never mock something you don't own (as you don't know the behavior of a class you don't own). This implies you should not mock doctrine, and so there is no way to change Doctrine.

**Comment by GrahamCampbell** [comment]
I agree with @stof. Mocking can easily be abused.

**Comment by Nek-** [comment]
I don't mock Doctrine :-) . I just don't want to have to change my code *for testing*.

Also there is a workaround with the `namespace Foo\Bar {}` notation: if you include that in your file, php will take the class inside this one instead of the on defined in the standard autoloader.

**Comment by stof** [comment]
@Nek- this still requires that the real class is not loaded in the same process. Class names are unique.

and for PHPunit or PhpSpec, it would mean you cannot test your class in this process if a previous test was mocking it (you would test the mock class), which makes things worse. This is why I said it makes huge assumptions on the way the mocking library will be used (making it totally unsuited for standalone mocking tools btw)

**Comment by Nek-** [comment]
No, here is the workaround:

```php
<?php
// src/Acme/Foo.php
namespace Acme;

final class Foo {
   public function something() {}
}
```

```php
<?php
// generated file
namespace Acme {

class Foo {
   public function something() {}
}

}
```

```php
require 'generatedFile.php';

class Mock extends Foo {}
```

**Comment by stof** [comment]
@Nek- this is **not** a workaround. As I said multiple time before, this means that your original class cannot be defined at all in the process anymore, and it will work only if the original class was never loaded before, which is very hard to implement

**Comment by Nek-** [comment]
I don't think it's that hard to guess if the class is used before. (that's a search over the spec folder)

But indeed if there is something like `__DIR__`, `__FILE__` used inside the class there will be a problem. But there is not so much differences with the original. (or maybe I still can't see what differences there is ?)

Maybe the functionnality can be implemented but only activated as option ? Or better: using the hhvm way and copying the class only when a special attribute is here on the mock (and the original class is final).

**Comment by stof** [comment]
@Nek- no, it is a search over the whole codebase, and anything running in the process (note that Prophecy is not even aware of a ``spec`` folder).

And as I said before, the reliable way to alter class definitions is to hook into the loading of files (to be able to do it at the right time), and this can be implemented fully outside Prophecy as it would be a totally independent tool (and it would work whatever your mocking tool btw).
And requiring a special attribute is an option. Just define a convention about the way to detect such classes in the source code.
but this tool removing the final keyword from class definition should be used carefully as it could hide issues with your codebase as I said previously.

**Comment by stof** [comment]
actually, not as I said previously as the comment was not yet posted but was waiting in the comment form in another tab. so here it goes:

and this also change things. If you have a class trying to extend your final class, your testsuite will pass if the mocking library redefined the class without the final keyword, because the testsuite will not test your real code but a different codebase now.

And the mocking library would have to redefine all classes before they are ever used in your testsuite, meaning that the only way is to redefine all final classes when loading them to be sure they can be mocked in the future.
and btw, if you really want to do such hack in your project, it is perfectly possible without any change in Prophecy (as far as Prophecy is concerned, ``Acme\Foo`` will not be final as you would load a class definition without the final keyword, and so it will be mockable). What you would have to do is register a stream wrapper which would alter the source code at the time the file is loaded (ask @nicolas-grekas about it, he did such things in the past, for things much more crazy than removing final keywords, like https://github.com/nicolas-grekas/Patchwork/blob/62ec83585ec22ccbd7efea60e4c70163a7905733/pieces/HTMLPurifier/class/FILTER.php#L12 for instance)

**Comment by TomasVotruba** [comment]
There is a new way to mock final classes with no problems:
https://phpfashion.com/how-to-mock-final-classes

**Comment by jakzal** [comment]
> I don't mock Doctrine :-) . I just don't want to have to change my code for testing.

Why do you feel you need to change your code for testing?

**Comment by FabianPiconeDev** [comment]
> Why do you feel you need to change your code for testing?
> 
> Instead of mocking the final class, I mock the interface it implements.

For people that dont use [dependency inversion](https://en.wikipedia.org/wiki/Dependency_inversion_principle), they have no interface to mock. They would have to add an interface to the final class and that would result a production code change.

**Comment by jakzal** [comment]
> For people that dont use dependency inversion, they have no interface to mock. 

I know it might be frustrating, but prophecy being an opinionated framework, doesn't aim to satisfy everyone. I expect dependency inversion falls into what prophecy's authors considered a good practice to follow.

**Comment by stof** [comment]
If you don't use DI, you cannot inject a mock, be it for an interface or for an actual class.

**Comment by jakzal** [comment]
> If you don't use DI, you cannot inject a mock, be it for an interface or for an actual class.

You don't have to follow Dependency Inversion, but still follow Dependency Injection.

**Comment by TomasVotruba** [comment]
The solution is much easier than we think: https://www.tomasvotruba.cz/blog/2019/03/28/how-to-mock-final-classes-in-phpunit


I use this approach regardless testing framework and it work great: https://github.com/dg/bypass-finals

**Comment by jakzal** [comment]
@TomasVotruba it's not a matter of how easy it is but whether it should be done at all. IMO it's an anti-pattern that doesn't belong to the core of a framework that claims to promote good coding practices. 

I can understand the urge to satisfy a wider user base though. It feels like an extension system would be useful to accommodate for custom use cases.

**Comment by TomasVotruba** [comment]
Sure, that's a tip for solution in a new package, regardless what other people prefer.

**Comment by FabianPiconeDev** [comment]
> There is a new way to mock final classes with no problems:
> https://phpfashion.com/how-to-mock-final-classes

However its solved, it should be solved in the testing framework, otherwise people have to remember “when i want to mock final classes, i need that other tool”. And they have to implement it separately which is a hurdle.

Maybe you can implement that in Prophecy with a pull request? Maybe as an option disabled by default to consider using dependency inversion. For people not using it, they can optionally bypass the finals. You can still provide a standalone package for other usages.

**Comment by ciaranmcnulty** [comment]
like Jakub said, mocking final classes is probably not a good idea (happy to continue debating that)

To be specific about Prophecy, we are not interested in supporting all possible uses. We are interested in supporting all good practices.

**Comment by Nek-** [comment]
Aren't final classes a good practice?

**Comment by stof** [comment]
@Nek- Mocking them is not. If your class is final, and you still need a mock, it probably means that the class is missing an interface that can be mocked instead.

**Comment by Nek-** [comment]
@stof what's the point with a using phpspec on an interface?

**Comment by Nek-** [comment]
Hummm. I believe sometimes in projects you just want to make stuff final but also do not want to add complexity by using an interface.

But well, I get that an interface fix the problem and for example in an open source project interfaces MUST be used. It's only about internal projects.

**Comment by FabianPiconeDev** [comment]
> Hummm. I believe sometimes in projects you just want to make stuff final but also do not want to add complexity by using an interface.
> 
> But well, I get that an interface fix the problem and for example in an open source project interfaces MUST be used. It's only about internal projects.

With those “sometimes” you facilitate class coupling. This complexity (which is only an abstraction per class) is designed to decouple classes. For every project, not specific for open source projects.

**Comment by Nek-** [comment]
Can somebody explain how API Platform can use prophecy to mock final classes? https://github.com/api-platform/core/blob/d83d50228e2686b7a2f76bfbadf2dec3fd83fdcd/tests/EventListener/ExceptionListenerTest.php#L38

**Comment by ciaranmcnulty** [comment]
@Nek- It was only made final in Symfony 5 https://github.com/symfony/symfony/commit/953057f28d82c06a73d393a2f64691f4a986d0a9#diff-0039df994b1b952fb5c9bf03294e7535

**Comment by Nek-** [comment]
@ciaranmcnulty yes, but API P tests passes with Sf 5.1: https://github.com/api-platform/core/runs/685864861

**Comment by drupol** [comment]
We are trying to implement tests for the PHPSpec extension [friends-of-phpspec/phpspec-code-coverage](https://packagist.org/packages/friends-of-phpspec/phpspec-code-coverage) in [this pr](https://github.com/friends-of-phpspec/phpspec-code-coverage/pull/39).

And we are blocked because the [CodeCoverage](https://github.com/sebastianbergmann/php-code-coverage/blob/master/src/CodeCoverage.php) class is set as final and we are unable to mock it in the tests.

Adding an interface to that class would help to mock it, as suggested by many people [here](https://github.com/phpspec/prophecy/issues/197) here.

I opened the [PR](https://github.com/sebastianbergmann/php-code-coverage/pull/833), let's see how it goes.

**Comment by stof** [comment]
@drupol a general rule is that you should not mock what you don't own (as you could not ensure that your mock definition will keep reflecting the actual behavior of the external library, and so your mock-based test does not actually ensure that your code works against the actual library). so I would not call that a use case here

**Comment by drupol** [comment]
> @drupol a general rule is that you should not mock what you don't own (as you could not ensure that your mock definition will keep reflecting the actual behavior of the external library, and so your mock-based test does not actually ensure that your code works against the actual library). so I would not call that a use case here

Fair enough ! I closed the phpunit PR in the meantime as well.

Thanks for your constructive answers all the time! Learning a lot from you!

**Comment by tuancode** [comment]
@Nek- You need to switch to the main branch to see the new unit test.
